drivers/video/gc2145.c

```diff
@@ -0,0 +1,1145 @@
+	 * just to be sure that the connection error is not caused by driver
+	 * itself.
+	 */
+	do {
+		ret = i2c_reg_write_byte_dt(spec, reg_addr, value);
+		if (ret < 0) {
+			LOG_ERR("failed to write 0x%x to 0x%x", value, reg_addr);
+			/* If writing failed wait 5ms before next attempt */
+			k_msleep(5);
+		}
+
+	} while ((tries--) || (ret != 0));
+
+	return ret;
+}
+
+static int gc2145_read_reg(const struct i2c_dt_spec *spec, uint8_t reg_addr, uint8_t *value)
+{
+	int ret = 0;
+	uint8_t tries = 3;
+
+	/**
+	 * It rarely happens that the camera does not respond with ACK signal.
+	 * In that case it usually responds on 2nd try but there is a 3rd one
+	 * just to be sure that the connection error is not caused by driver
+	 * itself.
+	 */
+	do {
+		ret = i2c_reg_read_byte_dt(spec, reg_addr, value);
+		if (ret < 0) {
+			LOG_ERR("failed to read 0x%x register", reg_addr);
+			/* If writing failed wait 5ms before next attempt */
+			k_msleep(5);
+		}
+
+	} while ((tries--) || (ret != 0));
+
+	return ret;
+}
+
+static int gc2145_write_all(const struct device *dev, const struct gc2145_reg *regs,
+			    uint16_t reg_num)
+{
+	uint16_t i = 0;
+	const struct gc2145_config *cfg = dev->config;
+
+	for (i = 0; i < reg_num; i++) {
+		int err;
+
+		err = gc2145_write_reg(&cfg->i2c, regs[i].addr, regs[i].value);
+		if (err) {
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int gc2145_soft_reset(const struct device *dev)
+{
+	int ret = 0;
+	const struct gc2145_config *cfg = dev->config;
+
+	/* Initiate system reset */
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_RESET, GC2145_REG_SW_RESET);
+
+	k_msleep(30);
+
+	return ret;
+}
+
+static int gc2145_set_ctrl_vflip(const struct device *dev, int enable)
+{
+	int ret;
+	const struct gc2145_config *cfg = dev->config;
+	uint8_t old_value = 0;
+
+	ret = gc2145_read_reg(&cfg->i2c, GC2145_REG_AMODE1, &old_value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set the vertical flip state */
+	return gc2145_write_reg(&cfg->i2c, GC2145_REG_AMODE1, (old_value & 0xFC) | (enable << 1));
+}
+
+static int gc2145_set_ctrl_hmirror(const struct device *dev, int enable)
+{
+	int ret;
+	const struct gc2145_config *cfg = dev->config;
+	uint8_t old_value = 0;
+
+	ret = gc2145_read_reg(&cfg->i2c, GC2145_REG_AMODE1, &old_value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set the horizontal mirror state */
+	return gc2145_write_reg(&cfg->i2c, GC2145_REG_AMODE1, (old_value & 0xFC) | enable);
+}
+
+static int gc2145_set_window(const struct device *dev, uint16_t reg, uint16_t x, uint16_t y,
+			     uint16_t w, uint16_t h)
+{
+	int ret;
+	const struct gc2145_config *cfg = dev->config;
+
+	ret = gc2145_write_reg(&cfg->i2c, 0xFE, GC2145_SET_P0_REGS);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Y/row offset */
+	ret = gc2145_write_reg(&cfg->i2c, reg++, y >> 8);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gc2145_write_reg(&cfg->i2c, reg++, y & 0xff);
+	if (ret < 0) {
+		return ret;
+	}
```

Comment:
do you need to write these one byte at a time? Also this API is really all over the place, does it make sense to call it with `reg` defined to anything other than `0x09`?

Reply:
Maybe `gc2145_write_all()` could be used for this? With a local array containing:
`uint8_t regs[] = { y >> 8, y & 0xff, x >> 8, x & 0xff, h >> 8, h & 0xff, w >> 8, w & 0xff };` for instance.

Reply:
`gc2145_write_all()` is a for loop writing them one by one, but then it would be possible to try to replace it to a batch write for all register loops rather than just one, i.e. to optimize startup time.

Reply:
(marked as resolved by @uLipe)
Note: the comments above were open suggestion: I have no problem with how the code is. :)